core/src/main/java/org/jruby/truffle/nodes/control/DoWhileNode.java

```diff
@@ -67,6 +65,7 @@ public boolean executeRepeating(VirtualFrame frame) {
                     break;
                 } catch (RedoException e) {
                     // Just continue in the while(true) loop.
+                    context.getSafepointManager().poll();
```

Comment:
We do not need a safepoint if the block is executed only once. In general, doing the safepoint only *after* the first iteration reduces the number of safepoints compiled/executed.